[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100856
* #100855

The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

fegin committed:
[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard

The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard"


The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard"


The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -612,38 +603,9 @@ def _sharded_pre_load_state_dict_hook(
-        tensors_to_flatten.append(tensor)
+        state_dict[fqn_from_global_root] = tensor
 
-    # Create a new flat_param from the loaded, non-sharded tensors.
```
I am guessing all of this logic got factored into the `_enter_unshard_params_ctx()`?

fegin replied:
Yes, the logic now is similar to full state_dict. So the main logic is to use `_enter_unshard_params_ctx()`.

fegin committed:
Update on "[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard"


The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict] Make sharded_state_dict work with composable fully_shard"


The current implementation of sharded_state_dict only works with wrapper based FSDP (both use_orig_params and not use_orig_params work) but not with fully_shard. This PR changes the implementation of sharded_state_dict when loading to fix the incompatibility.

Differential Revision: [D45626856](https://our.internmc.facebook.com/intern/diff/D45626856/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D45626856/)!

[ghstack-poisoned]
